### PR TITLE
RDKBWIFI-323: Policy fixes and changes for single AP Metrics Report

### PR DESCRIPTION
**Pull request by Nikita-Hakai**
- Policy fixes:
    - Added new custom policy and vendor tlvs to send the custom policy
    - Fixed duplicate policy sending and policy to send per device
    - For AP related policies, updated code to parse policy for all radios if CLI sends radio Mac as ff:ff:ff:ff:ff and controller internally considers it as all radio and formulates radio Mac in the radio metrics reporting accordingly.
    - policy ff:ff:ff:ff:ff fixed so that this parse is to parse for all radios
- AP metrics Report
    - Code is updated to formulate one single report containing all radio tlvs as per the policy - stress testing needed for corner cases
    - comment tests alarms
- Link Stats Alarm Report
    - encode received link stats alarm report
    - receive linkreport of sta, encode a json and publish to orch!

### CODE REVIEW

File: inc/dm_easy_mesh_ctrl.h [hunk 1]
```diff
@@ -557,7 +557,7 @@ class dm_easy_mesh_ctrl_t :
 	 *
 	 * @note Ensure that the event and command pointers are valid before calling this function.
 	 */
-	int analyze_m2_tx(em_bus_event_t *evt, em_cmd_t *cmd[]);
+	int analyze_m2_tx(em_bus_event_t *evt, em_cmd_t *cmd[], bool evt_status);
```
Comment (amarnathhullur) [inline inc/dm_easy_mesh_ctrl.h:561]: Add documentation for the third argument introduced.

File: inc/dm_easy_mesh_ctrl.h [hunk 2]
```diff
@@ -740,7 +740,7 @@ class dm_easy_mesh_ctrl_t :
 	 *
 	 * @note Ensure that the JSON object and key are valid before calling this function.
 	 */
-	int get_sta_config(cJSON *parent, char *key, em_get_sta_list_reason_t reason = em_get_sta_list_reason_none);
+	int get_sta_config(cJSON *parent, char *key, em_get_sta_list_reason_t reason = em_get_sta_list_reason_none, char *data = NULL);
```
Comment (amarnathhullur) [inline inc/dm_easy_mesh_ctrl.h:746]: Add documentation for the char* data pointer introduced

File: inc/em_base.h [hunk 3]
```diff
@@ -628,8 +628,10 @@ typedef enum {
     em_tlv_type_avail_spectrum_inquiry_reg = 0xe8,
     em_tlv_type_avail_spectrum_inquiry_rsp = 0xe9,
     em_tlv_type_vendor_sta_metrics = 0xf1,
-    em_tlv_vendor_plolicy_cfg = 0xf2,
-    em_tlv_type_vendor_operational_bss = 0xf3,
+    em_tlv_type_vendor_policy_cfg_sta_marker = 0xf2,
+    em_tlv_type_vendor_policy_cfg_alarm = 0xf3,
+    em_tlv_type_vendor_operational_bss = 0xf4,
+    em_tlv_type_link_stats_alarm_rprt = 0xf5,
 } em_tlv_type_t;
```
Comment (amarnathhullur) [inline inc/em_base.h:641]: All of these tlv_type_vendor should be made part of vendor_ext_attr_id_t and removed from here. 
The construction and handling should happen similar to that of PR#539. This will be in sync with spec for interoperability.

File: inc/em_base.h [hunk 4]
```diff
@@ -1105,6 +1107,23 @@ typedef struct {
     unsigned char reserved[2];
 } __attribute__((__packed__)) em_spatial_reuse_rprt_t;
 
+typedef struct {
+    float link_quality_score;
+    char reporting_time[32];
```
Comment (amarnathhullur) [inline inc/em_base.h]: Remove magic number

File: inc/em_base.h [hunk 5]
```diff
@@ -1105,6 +1107,23 @@ typedef struct {
...
+    int snr;
+    int per;
+    int phy;
```
Comment (amarnathhullur) [inline inc/em_base.h]: All the above parameters should be float similar to that of link_quality_score as per my understanding

File: inc/em_base.h [hunk 6]
```diff
@@ -1105,6 +1107,23 @@ typedef struct {
...
+} __attribute__((__packed__)) em_alarm_samples_t;
+
+typedef struct {
+    mac_addr_t sta_mac;
+    unsigned char reporting_timestamp[32];
```
Comment (amarnathhullur) [inline inc/em_base.h]: Remove magic number and replace with appropriate macro.

File: inc/em_base.h [hunk 7]
```diff
@@ -1105,6 +1107,23 @@ typedef struct {
...
+    float link_quality_threshold;
+    bool alarm_triggered;
+    int sample_count;
+    em_alarm_samples_t alarm_sample[3];
```
Comment (amarnathhullur) [inline inc/em_base.h]: Remove magic number and replace with appropriate macro. Is 3 enough?

File: inc/em_base.h [hunk 8]
```diff
@@ -1804,8 +1823,15 @@ typedef struct {
     unsigned char  reserved[20];
 } __attribute__((__packed__)) em_qos_mgmt_policy_t;
 
+typedef struct {
+    em_long_string_t collection_start_time;
```
Comment (amarnathhullur) [inline inc/em_base.h]: The size of em_long_string_t is 128 bytes, where as the reporting_timestamp above is 32 bytes. Shouldn't this be just em_string_t?

File: inc/em_base.h [hunk 9]
```diff
@@ -2320,6 +2347,8 @@ typedef struct {
     em_long_string_t    vendor_info[MAX_VENDOR_INFO];
 
     wifi_BeaconReport_t beacon_reports[EM_MAX_BEACON_REPORTS_PER_SCAN];
+    //link stats alram
```
Comment (amarnathhullur) [inline inc/em_base.h]: Typo in comment - alarm

File: inc/em_base.h [hunk 10]
```diff
@@ -3252,6 +3285,12 @@ typedef struct {
     em_ssid_info_traffic_separ_pol_t ssid_info[em_haul_type_max];
 } em_traffic_separation_policy_t;
 
+typedef struct {
+    em_long_string_t collection_start_time;
```
Comment (amarnathhullur) [inline inc/em_base.h:3288]: Change em_long_string_t appropriately

File: src/agent/dm_easy_mesh_agent.cpp [hunk 11]
```diff
@@ -857,52 +857,63 @@ int dm_easy_mesh_agent_t::analyze_ap_metrics_report(em_bus_event_t *evt, em_cmd_
     em_cmd_params_t *evt_param = NULL;
     unsigned int num = 0;
     mac_addr_str_t macstr;
-    cJSON *json, *emap_metrics_report, *param_obj;
+    cJSON *json, *emap_metrics_report, *param_obj, *arr_obj;
     int radio_index = 0;
     
-    printf("%s:%d: Enter\n", __func__, __LINE__);
+    em_printfout("Enter");
```
Comment (amarnathhullur) [inline src/agent/dm_easy_mesh_agent.cpp]: Do we really need this log?

**Review by amarnathhullur — CHANGES_REQUESTED** [review]
Changes related to inc files

File: src/agent/dm_easy_mesh_agent.cpp [hunk 12]
```diff
@@ -857,52 +857,63 @@ int dm_easy_mesh_agent_t::analyze_ap_metrics_report(em_bus_event_t *evt, em_cmd_
...
+
+        param_obj = cJSON_GetObjectItemCaseSensitive(arr_obj, "Radio Index");
+        if ((param_obj == NULL) || (cJSON_IsNumber(param_obj) == false) ) {
+            em_printfout("Unable to find scanner mac");
+            return 0;
+        }
+
+        radio_index = param_obj->valueint;
+        evt_param = &evt->params;
+        dm_easy_mesh_t::macbytes_to_string(get_radio_by_ref(radio_index).get_radio_interface_mac(), macstr);
```
Comment (amarnathhullur) [inline src/agent/dm_easy_mesh_agent.cpp]: if macstr is used only for printing, instead of using macbytes_to_string use 
util::mac_to_string(get_radio_by_ref(radio_index).get_radio_interface_mac()).c_str()

Comment (amarnathhullur) [inline src/agent/dm_easy_mesh_agent.cpp]: since macstr is used only for logging, use the function
util::mac_to_string(get_radio_by_ref(radio_index).get_radio_interface_mac()).c_str() for printing.

File: src/ctrl/tr_181/wfa_data_model/tr_181.cpp [hunk 13]
```diff
@@ -747,6 +754,85 @@ bus_error_t tr_181_t::affap_tget(char *event_name, raw_data_t *p_data, bus_user_
...
+{
+    p_data->data_type       = bus_data_type_string;
+    p_data->raw_data.bytes  = malloc(sizeof(mac_addr_str_t));
+    if (p_data->raw_data.bytes == NULL) {
+        em_printfout("Memory allocation is failed");
+        return bus_error_out_of_resources;
+    }
+    em_printfout(" get_node_sync: node mac len: %d", sizeof(mac_addr_str_t));
+
+    strncpy((char *)p_data->raw_data.bytes, (const char *)g_temp_node_mac, sizeof(g_temp_node_mac));
```
Comment (amarnathhullur) [inline src/ctrl/tr_181/wfa_data_model/tr_181.cpp]: would be better to have the 3rd parameter as sizeof(mac_addr_str_t) as that is what the raw_data_len is assigned to.

File: src/ctrl/dm_easy_mesh_ctrl.cpp [hunk 14]
```diff
@@ -234,6 +236,15 @@ int dm_easy_mesh_ctrl_t::analyze_m2_tx(em_bus_event_t *evt, em_cmd_t *pcmd[])
     tmp = pcmd[num];
     num++;
 
+    type = em_cmd_t::bus_2_cmd_type(evt->type);
+    em_printfout("type: %s, cmd ip: :%d", em_cmd_t::get_bus_event_type_str(evt->type), evt_status);
+    // If event status is true, override op to policy config to send only once if already active as multiple m2 tx events can be received
+    if (evt_status == true) {
```
Comment (amarnathhullur) [inline src/ctrl/dm_easy_mesh_ctrl.cpp:244]: Is evt_status becoming true only for the policy config?

File: src/ctrl/dm_easy_mesh_ctrl.cpp [hunk 15]
```diff
@@ -234,6 +236,15 @@ int dm_easy_mesh_ctrl_t::analyze_m2_tx(em_bus_event_t *evt, em_cmd_t *pcmd[])
...
+        desc.op = dm_orch_type_policy_cfg;
+        desc.submit = false;
+        pcmd[num - 1]->override_op(5, &desc);
```
Comment (amarnathhullur) [inline src/ctrl/dm_easy_mesh_ctrl.cpp]: Remove magic number

File: src/ctrl/dm_easy_mesh_ctrl.cpp [hunk 16]
```diff
@@ -234,6 +236,15 @@ int dm_easy_mesh_ctrl_t::analyze_m2_tx(em_bus_event_t *evt, em_cmd_t *pcmd[])
     tmp = pcmd[num];
     num++;
 
+    type = em_cmd_t::bus_2_cmd_type(evt->type);
+    em_printfout("type: %s, cmd ip: :%d", em_cmd_t::get_bus_event_type_str(evt->type), evt_status);
+    // If event status is true, override op to policy config to send only once if already active as multiple m2 tx events can be received
```
Comment (amarnathhullur) [inline src/ctrl/dm_easy_mesh_ctrl.cpp]: Add more document to indicate the intent behind this check

File: src/ctrl/dm_easy_mesh_ctrl.cpp [hunk 17]
```diff
@@ -605,47 +616,78 @@ int dm_easy_mesh_ctrl_t::analyze_dpp_start(em_bus_event_t *evt, em_cmd_t *cmd[])
...
-   		while ((pcmd[num] = tmp->clone_for_next()) != NULL) {
-       		tmp = pcmd[num];
-       		num++;
-   		}
+        dev_dm = get_data_model(GLOBAL_NET_ID, dm.m_device.m_device_info.intf.mac);
+        if (dev_dm != NULL) {
+            //compare if policy has changed for this device, create cmd only if a policy chnage is detected
+            for (int j = 0; j < dev_dm->get_num_policy(); j++) {
+                if ((dev_dm->m_policy[j] == dm.m_policy[j]) == false) {
+                    policy_changed++;
```
Comment (amarnathhullur) [inline src/ctrl/dm_easy_mesh_ctrl.cpp:654]: if policy_changed is incremented once, is there a need to go through all the remaining policy of the dev_dm? Maybe a break can be added to exit out of the for loop.

File: src/ctrl/dm_easy_mesh_ctrl.cpp [hunk 18]
```diff
@@ -605,47 +616,78 @@ int dm_easy_mesh_ctrl_t::analyze_dpp_start(em_bus_event_t *evt, em_cmd_t *cmd[])
...
-	return static_cast<int> (num);
+        if (policy_changed == 0) {
+            em_printfout("No Policy change detected for device with MAC: %s", mac_str);
+            i++;
+            continue;
+        } else {
+            em_printfout("Policy change detected for device with MAC: %s", mac_str);
+        }
+        radio = m_data_model_list.get_first_radio(dm.m_network.m_net_info.id, dm.m_device.m_device_info.intf.mac);
+        while (radio != NULL) {
```
Comment (amarnathhullur) [inline src/ctrl/dm_easy_mesh_ctrl.cpp:678]: I dont undersand the need to go through all the radio. Can you elaborate.

File: src/ctrl/dm_easy_mesh_ctrl.cpp [hunk 19]
```diff
@@ -605,47 +616,78 @@ int dm_easy_mesh_ctrl_t::analyze_dpp_start(em_bus_event_t *evt, em_cmd_t *cmd[])
...
+            radio = m_data_model_list.get_next_radio(dm.m_network.m_net_info.id, dm.m_device.m_device_info.intf.mac, radio);
+        }
+        dm.set_db_cfg_param(db_cfg_type_policy_list_update, "");
+        pcmd[num] = new em_cmd_set_policy_t(evt->params, dm);
+        tmp = pcmd[num];
+        num++;
+
+        em_printfout("Setting policy for Device number-%d with MAC: %s", i, mac_str);
+
+        i++;
```
Comment (amarnathhullur) [inline src/ctrl/dm_easy_mesh_ctrl.cpp:689]: shouldn't the policy_changed be reset to 0 for the next iteration?

File: src/ctrl/dm_easy_mesh_ctrl.cpp [hunk 20]
```diff
@@ -1550,6 +1629,11 @@ void dm_easy_mesh_ctrl_t::get_config(em_long_string_t net_id, em_subdoc_info_t *
         get_mld_config(parent, net_id);
     } else if (strncmp(subdoc->name, "WifiReset", strlen(subdoc->name)) == 0) {
         get_wifi_reset_config(parent, net_id);
+    } else if (strncmp(subdoc->name, "alarm_report", strlen(subdoc->name)) == 0) {
+        //em_printfout("Getting STAList for alarm report\n");
+        //get_sta_config(parent, net_id, em_get_sta_list_reason_alarm_report, subdoc->buff);
```
Comment (amarnathhullur) [inline src/ctrl/dm_easy_mesh_ctrl.cpp]: Why is this commented?

File: src/ctrl/em_ctrl.cpp [hunk 21]
```diff
@@ -207,13 +207,13 @@ void em_ctrl_t::handle_set_policy(em_bus_event_t *evt)
...
+        //m_ctrl_cmd->send_result(em_cmd_out_status_prev_cmd_in_progress);
     } else if ((num = m_data_model.analyze_set_policy(evt, pcmd)) == 0) {
-        m_ctrl_cmd->send_result(em_cmd_out_status_no_change);
+        //m_ctrl_cmd->send_result(em_cmd_out_status_no_change);
     } else if (m_orch->submit_commands(pcmd, static_cast<unsigned int> (num)) > 0) {
-        m_ctrl_cmd->send_result(em_cmd_out_status_success);
+        //m_ctrl_cmd->send_result(em_cmd_out_status_success);
     } else {
-        m_ctrl_cmd->send_result(em_cmd_out_status_not_ready);
+        //m_ctrl_cmd->send_result(em_cmd_out_status_not_ready);
```
Comment (amarnathhullur) [inline src/ctrl/em_ctrl.cpp]: Why are all the above commented?

File: src/dm/dm_easy_mesh.cpp [hunk 22]
```diff
@@ -931,21 +931,36 @@ int dm_easy_mesh_t::decode_config_set_policy(em_subdoc_info_t *subdoc, const cha
...
-        return EM_PARSE_ERR_GEN;
-    }
-
-    if ((net_obj = cJSON_GetObjectItem(net_obj, "Network")) == NULL) {
-        printf("%s:%d: Failed to parse: %s\n", __func__, __LINE__, subdoc->buff);
+    /* Look for 'Network' either under provided wrapper key or at top-level */
+    cJSON *wrapper_obj = NULL;
+    if (key != NULL && (wrapper_obj = cJSON_GetObjectItem(parent_obj, key)) != NULL) {
+        char *tmp = cJSON_PrintUnformatted(wrapper_obj);
+        printf("%s:%d: Found wrapper key '%s'. Wrapper: %s\n", __func__, __LINE__, key, tmp ? tmp : "<null>");
```
Comment (amarnathhullur) [inline src/dm/dm_easy_mesh.cpp]: change all printf to em_printfout especially in this file.

File: src/dm/dm_easy_mesh.cpp [hunk 23]
```diff
@@ -956,100 +971,127 @@ int dm_easy_mesh_t::decode_config_set_policy(em_subdoc_info_t *subdoc, const cha
...
-	if ((dev_obj = cJSON_GetArrayItem(dev_arr_obj, static_cast<int> (index))) == NULL) {
-        printf("%s:%d: Invalid input index: %d\n", __func__, __LINE__, index);
+        if ((dev_obj = cJSON_GetArrayItem(dev_arr_obj, static_cast<int> (index))) == NULL) {
+            printf("%s:%d: Invalid input index: %d\n", __func__, __LINE__, index);
+            cJSON_Delete(parent_obj);
+            return EM_PARSE_ERR_GEN;
+        }
+    } else if ((dev_obj = cJSON_GetObjectItem(net_obj, "Device")) != NULL) {
+        /* single Device object provided; only valid for index == 0 */
+        num_devices = 1;
```
Comment (amarnathhullur) [inline src/dm/dm_easy_mesh.cpp:1002]: why is there hardcoding of the value?

File: src/em/metrics/em_metrics.cpp [hunk 24]
```diff
@@ -899,14 +979,14 @@ int em_metrics_t::send_ap_metrics_response()
...
+        // //Radio Metrics TLV (17.2.60)
+        // tlv = reinterpret_cast<em_tlv_t *> (tmp);
+        // tlv->type = em_tlv_type_radio_metric;
+        // sz = create_radio_metrics_tlv(tlv->value);
+        // tlv->len =  htons(static_cast<unsigned short> (sz));
 
-        tmp += (sizeof(em_tlv_t) + static_cast<size_t> (sz));
-        len += (sizeof(em_tlv_t) + static_cast<size_t> (sz));
+        // tmp += (sizeof(em_tlv_t) + static_cast<size_t> (sz));
+        // len += (sizeof(em_tlv_t) + static_cast<size_t> (sz));
```
Comment (amarnathhullur) [inline src/em/metrics/em_metrics.cpp]: why is the above block fully commented?

**Review by amarnathhullur — CHANGES_REQUESTED** [review]

File: inc/em_base.h (hunk 7 above)
Comment (Nikita-Hakai) [reply inc/em_base.h]: Have been updated to 10, will also update in this review changes.

File: src/ctrl/dm_easy_mesh_ctrl.cpp (hunk 14 above)
Comment (Nikita-Hakai) [reply src/ctrl/dm_easy_mesh_ctrl.cpp:244]: It will become true for cmd type em_cmd_type_em_config associated to bus event em_bus_event_type_m2_tx, ie. during autoconfiguration renew.

File: src/ctrl/dm_easy_mesh_ctrl.cpp (hunk 18 above)
Comment (Nikita-Hakai) [reply src/ctrl/dm_easy_mesh_ctrl.cpp:678]: Since "dm" is a local dm and it wont have the data model's radio info of that device, I am manually copy radio info for this command, which is also later used in build_candidates to select the em.

File: src/ctrl/dm_easy_mesh_ctrl.cpp (hunk 19 above)
Comment (Nikita-Hakai) [reply src/ctrl/dm_easy_mesh_ctrl.cpp:689]: No, "i" is used to iterate over multiple device list from the json, for total number of "num_devices"

File: src/ctrl/dm_easy_mesh_ctrl.cpp (hunk 20 above)
Comment (Nikita-Hakai) [reply src/ctrl/dm_easy_mesh_ctrl.cpp]: It was a test code, cleaned up

File: src/dm/dm_easy_mesh.cpp (hunk 23 above)
Comment (Nikita-Hakai) [reply src/dm/dm_easy_mesh.cpp:1002]: So this decode parses single device(using for orchestrator) and also existing DeviceList, which is multiple devices(for CLI). Hence, if the Object name is Device, it would be a single device.

File: src/em/metrics/em_metrics.cpp (hunk 24 above)
Comment (Nikita-Hakai) [reply src/em/metrics/em_metrics.cpp]: its moved below

File: inc/em_base.h (hunk 3 above)
Comment (Nikita-Hakai) [reply inc/em_base.h:641]: Addressed sta_marker, policy cfg as vendor tlvs. Link report tlv will be handled in separate set of changes as that is already addressed there.

File: src/ctrl/dm_easy_mesh_ctrl.cpp [hunk 25]
```diff
@@ -1279,15 +1325,52 @@ int dm_easy_mesh_ctrl_t::get_policy_config(cJSON *parent, char *net_id)
...
+    em_device_info_t *info;
 
     net_obj = cJSON_AddObjectToObject(parent, "Network");
-    dm_network_list_t::get_config(net_obj, key, true);
+
+    if (reason == em_get_sta_list_reason_alarm_report) {
+        dm_network_list_t::get_config(net_obj, key, true);
+        dev_obj = cJSON_AddObjectToObject(net_obj, "Device");
+
+        dm_easy_mesh_t::macbytes_to_string(reinterpret_cast<unsigned char *>(data), mac_str);
```
Comment (amarnathhullur) [inline src/ctrl/dm_easy_mesh_ctrl.cpp:1345]: data is used without checking for NULL. Since the default parameter of this argument is NULL, NULL check should be addressed.

File: inc/em_base.h (hunk 3 above)
Comment (amarnathhullur) [reply inc/em_base.h:641]: Why is em_tlv_type_link_stats_alarm_rprt still present here? That needs to be added in vendor_ext_attr_id_t.

File: inc/em_base.h [hunk 26]
```diff
@@ -1963,6 +1988,10 @@ typedef enum {
     // Vendor Extension Attribute for Haul Type with data
     // of 1 byte having value corresponding to em_haul_type_t
     vendor_ext_attr_id_haul_type = 0x01,
+    vendor_ext_attr_id_policy_sta_marker,
+    vendor_ext_attr_id_policy_alarm,
```
Comment (amarnathhullur) [inline inc/em_base.h:1992]: Update for each of the attr_id the data associated with that.

File: inc/em_base.h [hunk 27]
```diff
@@ -1486,7 +1504,7 @@ typedef struct {
 
 typedef struct {
     unsigned char  vendor_oui[3];
-    unsigned short m_num;
+    unsigned char m_num;
```
Comment (amarnathhullur) [inline inc/em_base.h:1507]: This format of the next byte after oui to have number of vendor id should be also used for the attr_id_haul type.

File: inc/em_metrics.h [hunk 28]
```diff
@@ -491,7 +493,7 @@ class em_metrics_t {
 	 *
 	 * @note Ensure that the buffer is large enough to hold the TLV.
 	 */
-	short create_radio_metrics_tlv(unsigned char *buff);
+	short create_radio_metrics_tlv(unsigned char *buff, int index);
```
Comment (amarnathhullur) [inline inc/em_metrics.h:496]: Document the  index parameter

File: src/ctrl/em_ctrl.cpp [hunk 29]
```diff
@@ -631,6 +672,61 @@ void em_ctrl_t::publish_network_topology()
     } else {
         printf("%s:%d Topology publish fail\n",__func__, __LINE__);
     }
+
+#if 1
```
Comment (amarnathhullur) [inline src/ctrl/em_ctrl.cpp:676]: Why do we need this change?

File: src/ctrl/em_ctrl.cpp [hunk 30]
```diff
@@ -631,6 +672,61 @@ void em_ctrl_t::publish_network_topology()
...
+    //Test code here
+    // if (desc->bus_event_subs_fn(&m_bus_hdl, DEVICE_WIFI_DATAELEMENTS_NETWORK_NODE_CFG_POLICY, (void *)&em_agent_t::onewifi_cb, NULL, 0) != 0) {
+    //     printf("%s:%d bus get failed\n", __func__, __LINE__);
+    //     return;
+    // }
+    em_printfout("\n%s:%d TEST_POLICY_CFG start\n", __func__, __LINE__);
+
+    /* Read policy JSON from /nik/orch/policy.json and publish it */
+    const char *policy_path = "/nik/orch/policy.json";
```
Comment (amarnathhullur) [inline src/ctrl/em_ctrl.cpp:685]: the path is hardcoded if this change is required.

**Review by amarnathhullur — APPROVED** [review]

**Comment by amarnathhullur** [comment]
Few outstanding comments will be addressed in subsequent PR's
